### PR TITLE
Glide feature

### PR DESCRIPTION
**Pull request by eon704**
## Changes:
- Updated the Player State system, now all player behaviours are handled by states and are triggered within `FixedUpdate`.
- Updated the input gathering and updating of the state. Now `GatherInput` method is called at the beginning of `FixedUpdate`, and the read input is sent further down to states to act upon.
- Updated `GameManager` to expose events and properties, so that `GameUI` is able to subscribe to them and set itself up on `Start`. `GameManager` is no longer setting up UI.
- Added `GlideState`, now player can glide when falling, reducing the max fall speed.
- Updated `JumpingState`, now the player can perform a short jump by tapping the jump button. And perform a high jump by holding a button.
- Added coyote time to have a forgiving short time window for the player to jump even after becoming not grounded.
- Updated the jump button to change the icon when the gliding action becomes available.

## Recordings:

Short Jump:

https://github.com/KrillOrBeKrilled/GMTK-2023/assets/16372290/0b162bf3-84dd-4623-bb39-e8742b4c179d

High Jump:

https://github.com/KrillOrBeKrilled/GMTK-2023/assets/16372290/0da3e1f7-ad01-4120-93c0-cb090ff5a11e

Glide:

https://github.com/KrillOrBeKrilled/GMTK-2023/assets/16372290/4b0b6cd2-693f-474f-a1af-4cbaf76c35d0

### CODE REVIEW

**Review by Eryngii01 — CHANGES_REQUESTED** [review]
Love the reworking of the State pattern to make it more self-sustaining and consistent to encapsulate almost all player states (wondering if we should migrate the build function into this too)! Coyote time is very generous (perfect timing for me and probably for casual players) and a must-have for avoiding the pits!

The glide is very nice and fun to mess around with in the map and just needs a few adjustments:
- I think the jump felt a bit too springy (might have also felt more exaggerated with my tiny phone screen); when I tapped and let go as speedily as I could, the jump still reached an apex at halfway through the screen. However, the maximum jump height when holding the jump button was a perfect height for me, although if we later let the hen get damaged by ceiling traps it might be easy to accidentally jump into them. Maybe the range of jump forces can be widened by lowering the `JumpStartForce` and increasing the `JumpForceMaxDuration`?
- It took me a while to notice when the jump icon turns into the glide icon and I understood the change more so with the mechanics though I already knew what to expect. I think I'd prefer if the glide icon was shown immediately when the jump is executed for the first frame. Although I could understand why the change would occur at the apex of the jump to communicate the "long" jump...It's hard to say which is better in this case, but my confusion might also be from the button animations when you hold it down
- There's some lag between the jump movement and the `SetGroundedState` method that calls `TrapController.DisableTrapDeployment`, so the trap placement grid appears for the first few frames of the jump with the movement sounds (bugs me a little, but this could just be a nitpick!).

File: Assets/Code/Scripts/Core/Player/Input/PlayerStates/JumpingState.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            }
+
+            bool noMovement = Mathf.Approximately(moveInput, 0f);
+            if (!jumpPressed) {
+                State nextState = noMovement ? State.Idle : State.Moving;
+                this._playerController.ChangeState(nextState);
+                return;
+            }
+
+            bool shouldPlaySound = Time.time > this._jumpSoundPlayTime + SoundInterval;
```
Comment (Eryngii01) [inline Assets/Code/Scripts/Core/Player/Input/PlayerStates/JumpingState.cs:57]: I'm not too sure what this does if `OnEnter` already plays the jump SFX. Is there a specific edge case that this covers?

Comment (eon704) [reply Assets/Code/Scripts/Core/Player/Input/PlayerStates/JumpingState.cs:57]: Oh that, it was for playing the jump sound multiple times at intervals while in the `JumpState`. I forgot the interval value that I set doesn't allow for repeated sounds. I will tweak it a bit, and then we can hear the sound effect 2-3 times during the high jump 👌

**Comment by eon704** [comment]
Following the comments following changes were made:
- `JumpStartForce` reduced, `JumpForceMaxDuration ` increased.
- Adjusted the `SoundInterval`, now 1-3 jump sound effects will play depending on how long the jump button is pressed.
- Muted the default button press SFX on the **Jump** and **Build** buttons.

Not addressed & not fully fixed:
- Regarding the jump button changing its icon – it is a bit confusing what way would make more sense to players. Will do some research and address that in a separate PR. 
- Good catch on the `TrapController` lag! Thank you for catching that, totally missed it. I tried fixing it in code and by adjusting the grounded-hit-collider, it did help somewhat but in some cases you can still see the surveying happen after jumping 😕 Not sure how to fix that fully yet, I think should be okay for a separate PR 👌

File: Assets/Code/Scripts/Core/Player/Input/PlayerStates/JumpingState.cs (hunk 1 above)
Comment (Eryngii01) [reply Assets/Code/Scripts/Core/Player/Input/PlayerStates/JumpingState.cs:57]: Sounds good! I can hear it now, so it works perfectly 👍

**Review by Eryngii01 — APPROVED** [review]
The jump and glide are a lot more flexible now and work perfectly 👌

As for the jump and glides icons changing and the TrapController lag, it sounds good to address these at a later time! I can create tickets for them👌

**Comment by eon704** [comment]
Found a way to fix the `TrapController`!

**Pull request merged**